minicriu-client.c

```diff
@@ -102,12 +144,295 @@ static int writefile(const char *file, const char *buf, size_t len) {
+	ehdr.e_phnum = 0;
+	ehdr.e_shentsize = 0;
+	ehdr.e_shnum = 0;
+	ehdr.e_shstrndx = 0;
+
+	// Create PT_NOTE phdr
+	phdr[phnum].p_type = PT_NOTE;
+	phdr[phnum].p_flags = 0;
+	phdr[phnum].p_offset = 0;
+	phdr[phnum].p_vaddr = 0;
+	phdr[phnum].p_paddr = 0;
+	phdr[phnum].p_memsz = 0;
+	phdr[phnum].p_filesz = 0;
+	phdr[phnum++].p_align = 0;
+
+	FILE *proc_maps = fopen("/proc/self/maps", "r");
+	if (proc_maps == NULL) {
+		perror("Could not open maps file. Failed to create checkpoint.");
+		return 1;
+	}
+
+	// Initialize NT_FILE
+	nt_file.descsz = 0;
+	nt_file.descsz += sizeof(nt_file.count) + sizeof(nt_file.page_size);
+	nt_file.page_size = 0x1000;
+
+	// Create PT_LOAD and NT_FILE phdrs
+	char buffer[256];
+	while (fgets(buffer, sizeof(buffer), proc_maps)) {
+		void *addr_start, *addr_end;
+		char perms[8];
+		long ofs;
+		int name_start = 0;
+		int name_end = 0;
+
+		int res = sscanf(buffer, "%p-%p %7s %lx %*d:%*d %*lx %n%*[^\n]%n", &addr_start,
+			&addr_end, perms, &ofs, &name_start, &name_end);
+
+		if (res < 4) {
+			perror("sscanf. Failed to create checkpoint.");
+			fclose(proc_maps);
+			return 1;
+		}
+
+		// [vsyscall] is mapped to the same address in each process
+		if (!strncmp(buffer + name_start, "[vsyscall]", sizeof("[vsyscall]") - 1)) {
+			continue;
+		}
+
+		// Save mapped files
+		if (name_end > name_start && *(buffer + name_start) != '[') {
+			int count = nt_file.count;
+			nt_file.filemaps[count].start = (long int)addr_start;
+			nt_file.filemaps[count].end = (long int)addr_end;
+			nt_file.filemaps[count].fileofs = ofs / nt_file.page_size;
+			memcpy(nt_file.filepath[count], buffer + name_start, name_end - name_start);
+			nt_file.filepath[count][name_end - name_start] = '\0';
+			nt_file.descsz += sizeof(struct filemap) + name_end - name_start + 1;
+			nt_file.count++;
+		}
+
+		phdr[phnum].p_type = PT_LOAD;
+		phdr[phnum].p_flags = 0;
+		phdr[phnum].p_flags |= perms[0] == 'r' ? PF_R : 0;
+		phdr[phnum].p_flags |= perms[1] == 'w' ? PF_W : 0;
+		phdr[phnum].p_flags |= perms[2] == 'x' ? PF_X : 0;
+		phdr[phnum].p_offset = 0;
+		phdr[phnum].p_vaddr = (long unsigned int)addr_start;
+		phdr[phnum].p_paddr = 0;
+		phdr[phnum].p_memsz = addr_end - addr_start;
+		phdr[phnum].p_filesz = phdr[phnum].p_flags != 0 ? addr_end - addr_start : 0;
+		phdr[phnum++].p_align = 0x1000;
+	}
+
+	fclose(proc_maps);
+
+	// Updating headers
+	ehdr.e_phnum = phnum;
+	int prstatus_sz = mc_thread_counter * (sizeof(Elf64_Nhdr) + sizeof(struct elf_prstatus) + align_up(MC_OWNER_SIZE, 4));
+	int ntfile_sz = sizeof(Elf64_Nhdr) + align_up(nt_file.descsz, 4) + align_up(MC_OWNER_SIZE, 4);
+	phdr[0].p_filesz = prstatus_sz + ntfile_sz;
+	phdr[0].p_offset = sizeof(Elf64_Ehdr) + ehdr.e_phnum * ehdr.e_phentsize;
+	for (int i = 1; i < phnum; i++) {
+		phdr[i].p_offset = align_up(phdr[i - 1].p_offset + phdr[i - 1].p_filesz, phdr[i].p_align);
+	}
+
+	char filename[32];
+	sprintf(filename, "minicriu-core.%d", pid);
+	FILE *coreFile = fopen(filename, "w+");
```

Comment:
It's better to add a check that `coreFile` opened/created successfully.

Reply:
Fix in [1cf34b2](https://github.com/CRaC/minicriu/pull/9/commits/1cf34b287c2d9ca113eeee0a7925f7b388c124db)